tony-core/src/main/java/com/linkedin/tony/Utils.java

```diff
@@ -355,4 +353,20 @@ public static void cleanupHDFSPath(Configuration hdfsConf, Path path) {
       LOG.error("Failed to clean up HDFS path: " + path, e);
     }
   }
+
+  /**
+   * Copy a list of resources delimited by comma from hdfs to local directory.
+   * @param resources a list of resources to be copied, delimited by comma.
+   * @param localDir local directory that these resources will be copied to.
+   * @param hdfsConf the configuration file for HDFS.
+   * @throws IOException exception thrown during file copies.
+   */
+  public static void localizeResources(String resources, String localDir, Configuration hdfsConf) throws IOException {
```

Comment:
Instead of adding this method and having the TaskExecutors copy resources from HDFS to local disk, I think we should just rely on YARN's localization mechanism to localize resources. See `TonyClient.addLocalResources()`. I think we should just use that. If the resource is already on HDFS, then we don't need to do the copy from local to HDFS part; we just need to create a LocalResource for it.

P.S.: The variable names `scFileStatus` and `scRsrc` in TonyClient.addLocalResources are weird. What is `sc` supposed to stand for? Can you rename these variables?

Reply:
This is a problem with that, you can't localize a directory, that is why I piggyback on the copyToLocal method.

Reply:
Talked to @hungj , we should have two flags, one for the dependencies for each container - libraries like what we did for tony. This part will use `localResources` to benefit from the yarn file cache as well (however, this won't persist the file hierarchy)

The other one is a function pull data from hdfs to local, this is more like a convenient method for non-power users to pull data from hdfs to local

Reply:
localization will work but it shouldn't be used for localizing data, I believe the container will time out if localization has not finished by a timeout (10 min default), if data is localized and data grows, then more containers will timeout.

Reply:
Can we rename this function to something other than localizeResources? Don't want to confuse it with YARN localization

Reply:
Yah, the problem is for R script, I don't think they have a way to stream data from HDFS. Basically we need a way to workaround the issue to pull the data from HDFS for the local R binary to consume.

Plan to have a method `container_resources` for libraries and another `application_resources` method for the data stuff.

Reply:
Checked with the use cases, they normally just download the same csv files (~2GB each) and use R to process from different locations.

Updated plan it support only localized_resource array.

Reply:
Considering it is only 2GB, the time to localize should not pass 20s, which is far away from default timeout. We can revisit when user starts to abuse this flag.

Reply:
Looks like there are R packages that let you read data from HDFS. They should be able to use something like rhadoop to read directly from HDFS from their R script: https://community.hortonworks.com/articles/34047/read-from-hdfs-in-r-script.html

Reply:
ok, I'll only add localizable resources array then.